### PR TITLE
[JENKINS-36748] Add a test case and not process null specs

### PR DESCRIPTION
**Pull request by fbelzunc**
Add a test case per @KostyaSha and @jtnord request in #2463 and fix another NPE which is produced when `spec` is null.

https://issues.jenkins-ci.org/browse/JENKINS-36748

@reviewbybees

### CODE REVIEW

File: test/src/test/java/hudson/triggers/TriggerTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            super(cron);
+        }
+
+        @Override
+        public DescriptorImpl getDescriptor() {
+            return DESCRIPTOR;
+        }
+
+        @Override
+        public Object readResolve() {
```
Comment (jtnord) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: I take it this is needed to that `Trigger`s `readResolve()` is not called? - in which case possibly worth documenting so it is obvious, otherwise this looks like a no-op and someone could remove it in the future.

File: test/src/test/java/hudson/triggers/TriggerTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Rule
+    public JenkinsRule jenkinsRule = new JenkinsRule();
+
+    @Issue("JENKINS-36748")
+    @Test
+    public void testNoNPE() throws Exception {
+        jenkinsRule.getInstance().createProjectFromXML("whatever", new ByteArrayInputStream(("<project>\n  <builders/>\n  <publishers/>\n  <buildWrappers/>\n" + triggersSection() + "</project>").getBytes()));
+        final Calendar cal = new GregorianCalendar();
+        Trigger.checkTriggers(cal);
```
Comment (jtnord) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: it would be clearer to do
```java
try {
  Trigger.checkTriggers(cal);
} catch (Exception e) {
  assertNull("No exception should be thrown", e);
}
```

although what you have works.

File: test/src/test/java/hudson/triggers/TriggerTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package hudson.triggers;
+
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Felix Belzunce Arcos
```
Comment (jtnord) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: NIT: © should be the first thing in the file (it should come before the package)

File: test/src/test/java/hudson/triggers/TriggerTest.java (hunk 1 above)
Comment (KostyaSha) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: Nested Describables doesn't call `readResolve()`

File: test/src/test/java/hudson/triggers/TriggerTest.java (hunk 3 above)
Comment (KostyaSha) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: @jtnord i also tried to suggest to use correct emails to differentiate, but nobody cares ;)

**Comment by jglick** [comment]
:-1: as described in #2463.

File: test/src/test/java/hudson/triggers/TriggerTest.java (hunk 2 above)
Comment (jglick) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: I think it is fine as is.

File: test/src/test/java/hudson/triggers/TriggerTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        final Calendar cal = new GregorianCalendar();
+        Trigger.checkTriggers(cal);
+    }
+
+    private String triggersSection() {
+        String tagname = MockTrigger.class.getName().replace("$", "_-");
+        return "<triggers> \n <" + tagname + ">\n </" + tagname + ">\n  </triggers>\n";
+    }
+
+    public static class MockTrigger extends Trigger<Item> {
```
Comment (jglick) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: Overkill and unconvincing to boot. Simpler and more effective to use `@LocalData` to demonstrate that a project with an incomplete `config.xml` using a standard `Trigger` implementation does not break the system. (That is what the original bug report described, after all—the presence of Job DSL and GitHub PR plugins were inessential.)

Comment (jglick) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: So just use `TimerTrigger`, for example, but delete `<cron></cron>` from the XML after creating a sample project in the GUI.

Comment (fbelzunc) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: @jglick A standard trigger works - I can create the test... it fails when you create a trigger extending from `Trigger` and override `readResolve` without doing `super.readResolve`. See [this PR](https://github.com/jenkinsci/ghprb-plugin/pull/389).

Comment (fbelzunc) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: ok, did not see your last comment - I will try.

Comment (fbelzunc) [inline test/src/test/java/hudson/triggers/TriggerTest.java]: @jglick As pointer above the issue happens when `Trigger.readResolve` is overrided. I think the test is now good as it is.

File: core/src/main/java/hudson/triggers/Trigger.java [hunk 5]
```diff
@@ -90,7 +90,9 @@ public void start(J project, boolean newInstance) {
         this.job = project;
 
         try {// reparse the tabs with the job as the hash
-            this.tabs = CronTabList.create(spec, Hash.from(project.getFullName()));
+            if (spec != null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/triggers/Trigger.java]: 🐜 should be there a warning at least?

**Comment by oleg-nenashev** [comment]
ping @fbelzunc

**Comment by fbelzunc** [comment]
@oleg-nenashev Sorry - just added the `WARNING`. I hope it is fine now :-)

**Comment by oleg-nenashev** [comment]
Fine for me 🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done
Retriggering the build
